bfvmm/include/hve/arch/intel_x64/vcpu.h

```diff
@@ -322,19 +332,57 @@ class EXPORT_HVE vcpu : public bfvmm::vcpu
 
     /// @endcond
 
+public:
+
+    /// @cond
+
+    cpuid::delegator *cpuid_delegator() const
+    { return m_cpuid_delegator.get(); }
+
+    nmi::delegator *nmi_delegator() const
+    { return m_nmi_delegator.get(); }
+
+    cr::delegator *cr_delegator() const
+    { return m_cr_delegator.get(); }
+
+    invd::delegator *invd_delegator() const
+    { return m_invd_delegator.get(); }
+
+    msr::delegator *msr_delegator() const
+    { return m_msr_delegator.get(); }
+
+    unhandled_exit::delegator *unhandled_exit_delegator() const
+    { return m_unhandled_exit_delegator.get(); }
+
+    /// @endcond
+
 private:
 
     bool m_launched{false};
 
-    std::unique_ptr<exit_handler> m_exit_handler;
     std::unique_ptr<vmcs> m_vmcs;
     std::unique_ptr<vmx> m_vmx;
+
+    std::list<handler_delegate_t> m_init_delegates;
+    std::list<handler_delegate_t> m_fini_delegates;
+    std::array<handler_delegate_t, 128> m_exit_delegates;
```

Comment:
I can't think of a way to move the delegators out of the VCPU with this patterns. Reason being, you are registering exit handlers that need to be associated with a specific vcpu. The vcpu's "participation" in handling an exit is extremely minimal (see vcpu.cpp line 237). Maybe we can chat about this one next week if you have some other ideas?